src/MongoDB.Driver/Encryption/LibMongoCryptController.cs

```diff
@@ -0,0 +1,854 @@
+            }
+
+            var keyDocument = new RawBsonDocument(keyDocumentBytes);
+            await _keyVaultCollection.InsertOneAsync(keyDocument, cancellationToken: cancellationToken).ConfigureAwait(false);
+            return keyDocument["_id"].AsBsonBinaryData;
+        }
+
+        public BsonBinaryData DecryptField(BsonBinaryData wrappedBinaryValue, CancellationToken cancellationToken)
+        {
+            ThrowIfNotInitializedOrDisposed();
+
+            var wrappedValueBytes = GetWrappedValueBytes(wrappedBinaryValue);
+
+            try
+            {
+                using (var context = _cryptClient.StartExplicitDecryptionContext(wrappedValueBytes))
+                {
+                    return ProcessStates(context, databaseName: null, cancellationToken);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new MongoClientException(ex.Message, ex);
+            }
+        }
+
+        public async Task<BsonBinaryData> DecryptFieldAsync(BsonBinaryData wrappedBinaryValue, CancellationToken cancellationToken)
+        {
+            ThrowIfNotInitializedOrDisposed();
+
+            var wrappedValueBytes = GetWrappedValueBytes(wrappedBinaryValue);
+
+            try
+            {
+                using (var context = _cryptClient.StartExplicitDecryptionContext(wrappedValueBytes))
+                {
+                    return await ProcessStatesAsync(context, databaseName: null, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new MongoClientException(ex.Message, ex);
+            }
+        }
+
+        public byte[] DecryptFields(byte[] encryptedDocumentBytes, CancellationToken cancellationToken)
+        {
+            ThrowIfNotInitializedOrDisposed();
+
+            try
+            {
+                using (var context = _cryptClient.StartDecryptionContext(encryptedDocumentBytes))
+                {
+                    return ProcessStates(context, databaseName: null, cancellationToken);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new MongoClientException(ex.Message, ex);
+            }
+        }
+
+        public async Task<byte[]> DecryptFieldsAsync(byte[] encryptedDocumentBytes, CancellationToken cancellationToken)
+        {
+            ThrowIfNotInitializedOrDisposed();
+
+            try
+            {
+                using (var context = _cryptClient.StartDecryptionContext(encryptedDocumentBytes))
+                {
+                    return await ProcessStatesAsync(context, databaseName: null, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new MongoClientException(ex.Message, ex);
+            }
+        }
+
+        public void Dispose()
```

Comment:
With this implementation of `Dispose` it is not possible for multiple instances ` LibMongoCryptController` to share a single instance of `CryptClient`.

As soon as you `Dispose` any instance of `LibMongoCryptController` all other instances of `LibMongoCryptController` that share the same `CryptClient` now have a reference to a disposed (and unusable) `CryptClient`.

If you want multiple instances of `LibMongoCryptController` to share a single instance of `CryptClient` you will have to move "ownership" of the `CryptClient` instance to somewhere else like `Cluster`.

Reply:
Then why are you calling `_cryptClient.Dispose`?